actionview/CHANGELOG.md

```diff
@@ -1,3 +1,27 @@
+*   Allow templates to define which locals they accept.
+
+    For example:
+
+    ```erb
+    <%# def(message:) -%>
+    <%= message %>
+    ```
+
+    Default values can also be provided:
+
+    ```erb
+    <%# def(message: "Hello, world!") -%>
+    <%= message %>
+    ```
+
+    To disallow locals entirely, set `no_locals`:
+
+    ```erb
+    <%# no_locals %>
```

Comment:
This still uses the "locals" terminology, so I think it doesn't quite address https://github.com/rails/rails/pull/45602#discussion_r926913586.

I spoke with @byroot and asked him about several alternatives.  He said he is okay with `arguments`:

```erb
<%# arguments: (message:) %>
<%# arguments: () %>
```

He also said he's okay with including a colon, if it's suitable.  My preference would be to include a colon, to mirror e.g. `# frozen_string_literal: true`.

(@byroot If I'm wrong about any of that :point_up:, please correct me!)

Reply:
@jonathanhefner @byroot IMO using `locals` makes more sense for this API:

1) We're documenting it as `Explicit Locals`.
2) To a developer, this feature is affecting the behavior of the `locals` key passed to `render`.
3) This is _not_ a method signature in its purest sense, we only kwargs will work, not positional arguments. (Ideally we would parse the comment and raise if non-kwargs arguments were provided, but we do not at the moment)

Given all that and the discussion so far, I would propose we do:

```erb
<%# locals: (message:) %>
<%# locals: () %>
```

...and I should probably update the docs to mention that only kwargs are valid. 

_But_ I am also happy to implement as `arguments: ...`, just wanted to give my 2c before rewriting what I have here. What do you think?

Reply:
Ok, I'd like this to go forward, it's unfortunate that we're blocked on terminology. I'll get some opinion from other core members starting with John, we have a meeting soon so I'll bring up that PR, I should come back with a definite answer by the end of week (if I fail to do so please don't hesitate to ping me again)

Reply:
Sounds good! I don't feel the need to rush this, and I'm happy to go with whatever API y'all decided on ❤️ 

In the meantime I'll get this cleaned up so it's closer to being mergeable.

Reply:
Ok, so I discussed this with John last night, and he has no opinion on the naming, which leaves me with the decision -_-.

After sleeping on it, I think you convinced me that `locals:` is the right name. In absolute I'd still prefer something like `arguments` or `def`, but consistency is my number 1 concern, and these are passed with `locals: {}`, so any other naming would be inconsistent.

So go ahead with: `locals: (foo:)`. For no locals I think `locals: ()` is fine, we can always add an alias later if people are weirded out.

Let me know once you are done and I'll do the final review.

Reply:
> So go ahead with: locals: (foo:). For no locals I think locals: () is fine, we can always add an alias later if people are weirded out.

I had suggested overloading [local_assigns](https://github.com/rails/rails/pull/45602#issuecomment-1201829705) to be a Ruby method that's named to reflect the [ActionView::Template#local_assigns](https://edgeapi.rubyonrails.org/classes/ActionView/Template.html#method-i-local_assigns) method.

Had you all considered `local_assigns:` instead of `locals:` to mirror that interface?

Reply:
Thanks for the reply @byroot. My intention behind https://github.com/rails/rails/pull/45602#discussion_r935621446 wasn't to re-suggest a method, but to suggest magic comment syntax that relies on an existing and familiar method name. In this case, using `<%# local_assigns: (...) %>` instead of `<%# locals: (...) %>`.

Reply:
Right. 

I think I'm satisfied with `locals:`, that's what people use to pass these arguments to the template, hence why it was chosen.